src/main/java/fi/aalto/cs/apluscourses/utils/RemoteFileCache.java

```diff
@@ -79,14 +87,49 @@ public boolean isInvalid() {
       return !file.exists();
     }
 
-    public File getFile() throws IOException {
+    public File getFile(@NotNull Project project) throws IOException {
       synchronized (lock) {
         if (!isDownloaded) {
-          FileUtils.copyURLToFile(new URL(url), file, timeout, timeout);
+          copyUrlToFile(new URL(url), file, timeout, timeout, project);
           isDownloaded = true;
         }
         return file;
       }
     }
   }
+
+  private static void copyUrlToFile(@NotNull URL source,
+                                    @NotNull File destination,
+                                    int connectionTimeout,
+                                    final int readTimeout,
+                                    @NotNull Project project) throws IOException {
+    var connection = source.openConnection();
+    connection.setConnectTimeout(connectionTimeout);
+    connection.setReadTimeout(readTimeout);
+    var length = connection.getContentLengthLong();
+    var stream = connection.getInputStream();
+    var out = openOutputStream(destination);
+    copyLarge(stream, out, new byte[DEFAULT_BUFFER_SIZE], length, source, project);
+  }
+
+  private static void copyLarge(@NotNull InputStream input,
+                                @NotNull OutputStream output,
+                                byte @NotNull [] buffer,
+                                long length,
+                                @NotNull URL url,
+                                @NotNull Project project)
+      throws IOException {
+    var progress = PluginSettings
+        .getInstance()
+        .getMainViewModel(project)
+        .progressViewModel
+        .start((int) length / 1024,
+            getAndReplaceText("ui.ProgressBarView.downloading", url.getFile()), false);
+    int n;
+    while (EOF != (n = input.read(buffer))) {
+      output.write(buffer, 0, n);
+      progress.incrementBy(n / 1024);
```

Comment:
If the number of read bytes is smaller than 1024, the progress bar won't move - fix pls